Add network profiler response image preview

bkonyi opened the pull request:
Preview:

![image](https://user-images.githubusercontent.com/24210656/115435312-8d374580-a1be-11eb-86b1-c1c36cbae23e.png)

Fixes https://github.com/flutter/devtools/issues/2603

kenzieschmoll commented on packages/devtools_app/lib/src/network/network_request_inspector_views.dart:
```diff
@@ -151,6 +160,92 @@ class HttpResponseView extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    final img = image.decodeImage(data.encodedResponse);
+    // Format should be of the form: [image/$format]
```
can you use quotes in this comment instead? I was (may still be) confused on whether the brackets are part of the expected form value

bkonyi replied:
The brackets were part of the expected format (e.g., "[image/$format]"), but I've just removed it.

kenzieschmoll commented on packages/devtools_app/lib/src/network/network_request_inspector_views.dart:
```diff
@@ -151,6 +160,92 @@ class HttpResponseView extends StatelessWidget {
+      padding: _rowPadding,
+      child: Row(
+        crossAxisAlignment: CrossAxisAlignment.start,
+        children: [
```
if you use `mainAxisSize: MainAxisSize.min` I think you should be able to get rid of the constraints that are passed around everywhere and remove the wrapping container

kenzieschmoll commented on packages/devtools_app/lib/src/network/network_request_inspector_views.dart:
```diff
@@ -151,6 +160,92 @@ class HttpResponseView extends StatelessWidget {
+  Widget _buildRow(
+    BuildContext context,
+    String key,
+    dynamic value,
```
can this be a String instead of dynamic?

kenzieschmoll commented on packages/devtools_app/lib/src/network/network_request_inspector_views.dart:
```diff
@@ -151,6 +160,92 @@ class HttpResponseView extends StatelessWidget {
+            child: Text(
+              value,
+              overflow: TextOverflow.ellipsis,
+              maxLines: 5,
```
Is this necessary? the values passed to _buildRow above seem relatively small and even if they weren't, would we want to cut the data off?

bkonyi replied:
Probably not. Not sure how this got there in the first place... I just popped this old change off my stash, so my memory's a bit fuzzy.

bkonyi commented:
Addressed comments. Looks like we were processing the mime incorrectly before, so now `package:mime` is giving us `jpe` for `image/jpeg` instead of `jpg` (see below). Just figured it'd be worth pointing out.

<img width="1680" alt="Screen Shot 2021-04-20 at 10 30 40 AM" src="https://user-images.githubusercontent.com/24210656/115440017-1e5ceb00-a1c4-11eb-96f5-5a13d5a7b0f4.png">

kenzieschmoll commented:
> now package:mime is giving us jpe for image/jpeg instead of jpg

package:mime has multiple abbreviations for some types. See all the options for image/jpeg here: https://github.com/dart-lang/mime/blob/master/lib/src/default_extension_map.dart#L363-L365.

kenzieschmoll commented on packages/devtools_app/lib/src/network/network_request_inspector_views.dart:
```diff
@@ -151,6 +160,86 @@ class HttpResponseView extends StatelessWidget {
+            ),
+          ],
+        ),
+        LayoutBuilder(builder: (context, constraints) {
```
we aren't using the constraints anymore so we should be able to remove this layout builder

kenzieschmoll commented on packages/devtools_app/lib/src/network/network_request_inspector_views.dart:
```diff
@@ -151,6 +160,86 @@ class HttpResponseView extends StatelessWidget {
+    String key,
+    String value,
+  ) {
+    return Container(
```
change to Padding widget since padding is the only property being used from Container

bkonyi commented:
> > now package:mime is giving us jpe for image/jpeg instead of jpg
> 
> package:mime has multiple abbreviations for some types. See all the options for image/jpeg here: https://github.com/dart-lang/mime/blob/master/lib/src/default_extension_map.dart#L363-L365.

Right, but since there's no option there to actually select `jpeg`, do we care enough to work around it? `package:mime` will always return `jpe` for `image/jpeg` when using `extensionFromMime`.

kenzieschmoll commented:
There's a TODO there to consider a workaround for html as well - extensionFromMime returns htm. I'm fine with us adding a  workaround for common types we expect users to encounter.

bkonyi commented:
> There's a TODO there to consider a workaround for html as well - extensionFromMime returns htm. I'm fine with us adding a workaround for common types we expect users to encounter.

Done. Not sure where the TODO is though.

kenzieschmoll commented on packages/devtools_app/lib/src/http/http_request_data.dart:
```diff
@@ -94,11 +95,28 @@ abstract class HttpRequestData extends NetworkRequest {
+    if (mime.endsWith(']')) {
+      mime = mime.substring(0, mime.length - 1);
+    }
     // TODO(kenz): consider special casing some extensions. For example,
```
TODO I'm referring to is right here :)

bkonyi replied:
I'm blind... :-) Added handling for `text/plain` while I'm here.

kenzieschmoll commented:
![lgtm_army](https://user-images.githubusercontent.com/43759233/115456855-81587d00-a1d8-11eb-93e7-ee140de5d148.png)

bkonyi commented:
Merging on flaky red.
